src/locales.cc

```diff
@@ -32,303 +32,214 @@
 
-  // translators: header of table column - S is for 'Status' (package installed or not)
-  th << _("S");
-  // translators: header of table column - the name of the package
-  th << _("Name");
-  // translators: header of table column - the package summary
-  th << _("Description");
+  private:
+    unsigned _state = 0;
+  };
 
-  tbl << th;
+  using LocaleStateMap = std::map<Locale,LocaleState>;
 
-  for_( it, myLocale.selectableBegin(), myLocale.selectableEnd() )
+  /** 	lloc helper building the set of locales to show */
+  inline LocaleStateMap argsToLocaleSet( const std::vector<std::string> & localeArgs_r, bool showAll_r )
   {
-    TableRow tr(3);
-    zypp::ui::Status status = (*it)->status();
+    LocaleStateMap ret;
+    bool mustFixState = true;	// If locales a added in No state and later checked if Requested/Fallback
 
-    if ( status == zypp::ui::S_KeepInstalled || status == zypp::ui::S_Protected )
-      tr << "i";
-    else
-      tr << " ";
-    tr << (*it)->name();
-    tr << (*it)->theObj()->summary();
-
-    tbl << tr;
-  }
-
-  tbl.sort(1);
-
-  cout << tbl;
-}
-
-static zypp::LocaleSet relevantLocales( const std::vector<std::string> &localeArgs, bool relaxed )
-{
-  const zypp::LocaleSet & availableLocales( God->pool().getAvailableLocales() );
-
-  zypp::LocaleSet resultSet;
-
-  for_( it, availableLocales.begin(), availableLocales.end() )
-  {
-    if ( localeArgs.empty() )
+    if ( showAll_r )
     {
-      // without argument take only requested locales
-      if ( God->pool().isRequestedLocale(*it) )
+      for ( Locale rloc : God->target()->requestedLocales() )
+	ret[rloc].tagRequested();
+      for ( const auto & avloc : God->pool().getAvailableLocales() )
+	ret[avloc];	// state will be adjusted later
+    }
+    else if ( ! localeArgs_r.empty() )
+    {
+      // Search patterns are matched against available locales.
+      // Exact locales are added even if not in available locales.
+      const auto & av { God->pool().getAvailableLocales() };
+      for ( const auto & larg : localeArgs_r )
       {
-        resultSet.insert(*it);
+	std::string::size_type suffix = 0;
+	bool includeNoCountryCode = false;
+	if ( str::endsWith( larg, "_" ) )	// Suffix _ (_*) require a (not empty) country code
+	  suffix = 1;
+	else if ( str::endsWith( larg, "_*" ) )	// --"--  _* as convenience for the above
+	  suffix = 2;
+	else if ( str::endsWith( larg, "*" ) )	// Suffix * (except _*) also matches an empty country code
+	{
+	  suffix = 1;
+	  includeNoCountryCode = true;
+	}
+
+	if ( suffix )
+	{
+	  const std::string & stem { larg.substr( 0, larg.size() - suffix ) };
+	  for ( const auto & avloc : av )
+	  {
+	    if ( avloc.language().code() == stem && ( includeNoCountryCode || avloc.country().code() != CountryCode::noCode ) )
+	      ret[avloc];	// state will be adjusted later
+	  }
+	}
+	else	// Exact locale arg
+	  ret[Locale(larg)];	// state will be adjusted later
```

Comment:
We don't know valid/invalid locales. If you check `zypper lloc -a` you will find various flavors (kmr_Latn_SY, ca_valencia, ca_ES_valencia, en-gb,...). The Locale class is intended to be used as `language_COUNTRY` but in case of doubt it's just a string and able to match anything a package maintainer may invent.